### PR TITLE
fix(Rendering): more performance improvements

### PR DESCRIPTION
**Pull request by martinken**
A fair number of calls were being done that didn't
need to be done. Also some VAO rebuilding was always
happening when it should not have.

### CODE REVIEW

**Comment by martinken** [comment]
@jourdain  You might want to take a look at this one. Lots of crazy here. At some point I want to change the getArray(true) stuff to a different name. Maybe getArrayByReference to make it more clear and clean.

**Comment by jourdain** [comment]
I like the getReference approach as a fast path for everything. I guess we can talk about it tomorrow if you want.

**Comment by jourdain** [comment]
BTW, if I think it's a good first step should I merge?

File: Sources/Common/DataModel/DataSetAttributes/index.js [hunk 1]
```diff
@@ -131,7 +131,8 @@ function vtkDataSetAttributes(publicAPI, model) {
   };
 
   attrTypes.forEach((value) => {
-    publicAPI[`get${value}`] = () => publicAPI.getArrayByIndex(model[`active${value}`]);
+    const activeVal = `active${value}`;
+    publicAPI[`get${value}`] = () => publicAPI.getArrayByIndex(model[activeVal]);
```
Comment (jourdain) [inline Sources/Common/DataModel/DataSetAttributes/index.js]: Love that closure optimization...

File: Sources/Rendering/Core/Renderer/index.js [hunk 2]
```diff
@@ -57,8 +57,8 @@ function vtkRenderer(publicAPI, model) {
         // vtkLight::SetLightTypeToSceneLight()
       } else if (light.lightTypeIsHeadLight()) {
         // update position and orientation of light to match camera.
-        light.setPosition(camera.getPosition());
-        light.setFocalPoint(camera.getFocalPoint());
+        light.setPosition(camera.getPosition(true));
```
Comment (jourdain) [inline Sources/Rendering/Core/Renderer/index.js]: Should we get a fast path for both the set and get?

File: Sources/Rendering/Core/Viewport/index.js [hunk 3]
```diff
@@ -36,9 +36,11 @@ function vtkViewport(publicAPI, model) {
 
   // this method get all the props including any nested props
   publicAPI.getViewPropsWithNestedProps = () => {
-    model.allprops = [];
+    model.allprops = model.props;
     model.props.forEach((prop) => {
-      model.allprops = model.allprops.concat(prop.getProps());
+      if (prop.getNestedProps()) {
+        model.allprops = model.allprops.concat(prop.getProps());
+      }
```
Comment (jourdain) [inline Sources/Rendering/Core/Viewport/index.js]: May need some talk...

File: Sources/Rendering/OpenGL/Actor/index.js [hunk 4]
```diff
@@ -93,11 +107,10 @@ function vtkOpenGLActor(publicAPI, model) {
...
       publicAPI.activateTextures();
-    } else {
-      // deactivate textures
-      model.activeTextures.forEach((child) => {
-        child.deactivate();
-      });
+    } else if (model.activeTextures) {
+      for (let index = 0; index < model.activeTextures.length; index++) {
+        model.activeTextures[index].deactivate();
+      }
```
Comment (jourdain) [inline Sources/Rendering/OpenGL/Actor/index.js]: Did you noticed that the `forEach` were bad compare to a simple for loop?

Another approach for fast looping is that:

```
let count = model.activeTextures.length;
while(count--) {
  model.activeTextures[count].deactivate();
}
```

File: Sources/Rendering/OpenGL/PolyDataMapper/index.js [hunk 5]
```diff
@@ -1102,30 +1111,31 @@ function vtkOpenGLPolyDataMapper(publicAPI, model) {
     // bind some light settings
     let numberOfLights = 0;
 
-    const lights = ren.getLights();
-    Object.keys(lights).map(key => lights[key]).forEach((light) => {
+    const lights = ren.getLights(true);
+    for (let index = 0; index < lights.length; ++index) {
```
Comment (jourdain) [inline Sources/Rendering/OpenGL/PolyDataMapper/index.js]: Are Lights now an array instead of an object?

File: Sources/macro.js [hunk 6]
```diff
@@ -169,6 +169,8 @@ export function obj(publicAPI = {}, model = {}) {
     return subset;
   };
 
+  publicAPI.getOneProperty = val => model[val];
```
Comment (jourdain) [inline Sources/macro.js]: Could it be replaced with the getReference macro? (I've added it while reviewing your code and reading your comment)

**Comment by jourdain** [comment]
@martinken you may want to remove my last commit if you've tested and noticed that the forEach are slower or allocate more memory for the GC. Otherwise, theoretically they should be optimized by the JIT.

Another approach to speedup loops is the `while(count--){}` option.

**Comment by jourdain** [comment]
Moreover, we should talk about `getOneProperty`. ;-)

**Comment by jourdain** [comment]
My last commit is to support something like that:

```
light.setPositionFrom(camera.getPositionByReference());
light.setFocalPointFrom(camera.getFocalPointByReference());
light.modified(camera.getMTime());
```

**Comment by jourdain** [comment]
Hum... The tests are failing on my machine.

File: Sources/macro.js (hunk 6 above)
Comment (jourdain) [reply Sources/macro.js]: I guess in that case it will just be a naming thing.

File: Sources/Rendering/Core/Renderer/index.js (hunk 2 above)
Comment (jourdain) [reply Sources/Rendering/Core/Renderer/index.js]: See my last comment at the bottom of that chat...

File: Sources/macro.js (hunk 6 above)
Comment (martinken) [reply Sources/macro.js]: Yes I think the implementation of getReference would work for this case.

**Comment by martinken** [comment]
OK can you give that a shot and see if it passes for you?

**Comment by jourdain** [comment]
I've tried on my `day` Mac and it is working but I have a build error/warning:

```
/Users/seb/Documents/code/Web/vtk-js/Sources/Rendering/OpenGL/VertexArrayObject/index.js
  229:11  error  'gl' is assigned a value but never used  no-unused-vars
```

I'll checkout the old code to make sure that I can reproduce the issue on my `day` Mac. Otherwise you will have to wait for me to try on my `evening` Mac. ;-)

**Pull request merged**